### PR TITLE
ci: add tests for migration to CiliumEndpointSlice

### LINKED ISSUE #32057: CI coverage for migration path from CiliumEndpoint to CiliumEndpointSlice
Add CI coverage for migration from CiliumEndpoint to CiliumEndpointSlice for feature graduation to stable, associated with issue #31904.

### PR DESCRIPTION
**Pull request by jshr-w**
This PR adds CI to validate the migration path from CiliumEndpoint to CiliumEndpointSlice, by checking that the migration does not affect connectivity on a migrated cluster.

Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Fixes: #32057

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

### CODE REVIEW

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [inline .github/workflows/tests-ces-migrate.yaml]: Do we need a kernel version matrix in order to test CES migration? Those seem like different dimensions, such that the kernel version on the nodes should not impact the migration.

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [inline .github/workflows/tests-ces-migrate.yaml]: Subject to the answer to my first question, if we don't need different kernels then we may not need LVH VMs either, and may be able to greatly simplify the test. For instance if this follows something closer to the smoke tests instead, then the test would be simpler, more efficient, and may not require special secrets or test trigger access (?)

Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: Thanks for the suggestion! It seems like the kernel version matrix is irrelevant to this test, so I am adjusting it to be more similar to the smoke tests and not use the LVH VMs.

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [inline .github/workflows/tests-ces-migrate.yaml]: Since the conn-disrupt action you're using already deploys pods, I think you can remove this step

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [inline .github/workflows/tests-ces-migrate.yaml]: Non-blocking: you could use the [cilium-config action](https://github.com/cilium/cilium/blob/main/.github/actions/cilium-config/action.yml) here instead. IMO the setup is a bit cleaner.

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [inline .github/workflows/tests-ces-migrate.yaml]: nit
```suggestion
          tunnel: 'disabled'
```
You'd need to modify your `vars` step below, or if you switch to the `cilium-config` action you can just provide the value (ex: https://github.com/cilium/cilium/blob/57db22b2029cd9b725555e0971d9a9f36aad9540/.github/workflows/conformance-e2e.yaml#L304)

**Review by thorn3r — CHANGES_REQUESTED** [review]
Overall this looks great, thanks for taking this on! I left a few in-line suggestions. You may also want to add steps for [merging the sysdumps together](https://github.com/cilium/cilium/blob/57db22b2029cd9b725555e0971d9a9f36aad9540/.github/workflows/conformance-e2e.yaml#L427) and [updating the commit status](https://github.com/cilium/cilium/blob/57db22b2029cd9b725555e0971d9a9f36aad9540/.github/workflows/conformance-e2e.yaml#L450) (I believe this is needed in order to prevent `skipped` from reporting as failed).
Could you also squash your commits?

**Comment by thorn3r** [comment]
Just realized I jumped the gun a bit and this is still marked as draft :sweat_smile: Feel free to retag me for review when it's ready

**Comment by jshr-w** [comment]
> Just realized I jumped the gun a bit and this is still marked as draft 😅 Feel free to retag me for review when it's ready

No problem! Thank you for your feedback, I'll go through them and update accordingly :)

**Comment by jshr-w** [comment]
The connectivity test failures for client-egress-l7/pod-to-world/http-to-one.one.one.one, client-egress-l7-named-port/pod-to-world/http-to-one.one.one.one, pod-to-ingress-service/pod-to-ingress-service observed seem to be related to the Cilium Agent restart, rather than migration to CES. -> #32611

**Comment by jshr-w** [comment]
Hi @thorn3r, I think I'm ready for a re-review! :)

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [inline .github/workflows/tests-ces-migrate.yaml]: I think we should use workflow dispatch + ariane for triggering this workflow, instead of on pull request. This would make sure it only runs after a trigger command is issued (i.e. `/test`) and would allow directly testing changes to the workflow in a PR created from a branch.

Here's an example from another workflow: https://github.com/cilium/cilium/blob/834c188c8fcd5c776a87adba0b20d400215e06db/.github/workflows/conformance-clustermesh.yaml#L5
You'll want to add it to the [ariane config](https://github.com/cilium/cilium/blob/main/.github/ariane-config.yaml) as well

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [inline .github/workflows/tests-ces-migrate.yaml]:
```suggestion
  push:
    branches:
      - main
      - ft/main/**
    paths-ignore:
      - 'Documentation/**'
      - 'test/**'
```
If you add the paths to ignore here and use ariane for PR trigger, you can drop the `check_changes` step below

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [inline .github/workflows/tests-ces-migrate.yaml:132]: I don't think this is needed for sysdump to capture flows

**Review by thorn3r — COMMENTED** [review]
Nice work :sunglasses: I left a couple small suggestions, but other than that it LGTM

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [inline .github/workflows/tests-ces-migrate.yaml]: Why are these different configurations needed to test migration?

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml:132]: Looks like this was copied over from smoke tests via 0cbb855a58ffb, context here: https://github.com/cilium/cilium/pull/16968#issuecomment-885964109 . The comment and this workaround are a few years old, so it's possible that it's no longer necessary, but further investigation would be required to establish that.

As long as we have other examples in the tree doing the same thing, I don't see the harm in keeping this the same as the others. If we don't need this, we can always come back and remove it from here and the other workflows that do the same thing.

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml]: Interesting comment, I think it's worth discussing and coming to an agreement on the right times to use Ariane. I've typically thought of it as a way to protect against abuse of expensive/slow cloud resources. Given that this PR doesn't allocate external cloud resources, this aspect is not a concern.

> This would make sure it only runs after a trigger command is issued

Is this a feature or a hinderance? We have several smoke tests that don't require a trigger command.

> would allow directly testing changes to the workflow in a PR created from a branch

This is already working with this PR from an external contributor. The workflow seems to be running fine directly from the PR, see results here: https://github.com/cilium/cilium/actions/runs/9136043125/job/25124317623?pr=32268 . On the contrary, I think that if we switched this to Ariane, it would make it more difficult for contributors outside the Cilium organization to develop the tests.

File: .github/workflows/tests-ces-migrate.yaml
Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: I remember in a previous discussion we said that the matrix is not necessary, but we could consider having these cases -- I added it as a 'nice-to-have' but I can defer to your perspectives on whether it should be removed or there is any reason to keep (cc: @marseel)

Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml]: I don't see the value in running the tests multiple times with different configurations if those configurations are unrelated to the functionality of the test. On the other hand, if you know that the target production code is impacted by specific agent configurations, then I'd encourage putting those options into the matrix.

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [reply .github/workflows/tests-ces-migrate.yaml]: Interesting rebuttal :grin: All good points and agreed, I think most of my opinions are based on my own assumptions. Sounds like we might not want to use Ariane here.

> I've typically thought of it as a way to protect against abuse

My POV on this was more about wasting resources than abusing, but this could be my own bias. I use CI to smoke test draft PRs fairly often, it's usually inevitable that I forget to run one a make target and there's a lingering typo or a change to a helm value. I could be an outlier on this one, though. I assumed we were trying to move E2E/long-running tests mostly to Ariane.

> The workflow seems to be running fine directly from the PR

Yes, I think I may have been conflating this with workflows we had that ran on `pull_request_target`

File: .github/workflows/tests-ces-migrate.yaml
Comment (marseel) [reply .github/workflows/tests-ces-migrate.yaml]: I remember I said it's probably worth testing two configurations: native-routing and tunneling.
I can imagine the following cases:
- in native-routing, both agents derive identity information from ipcache, one from CEP, one from CES during the upgrade
- in tunneling, identity information is passed through the tunnel, could be derived either from CES or CEP and received on the other side that has different mode during the upgrade.

For `endpoint-routes` I think we can remove it. 
WDYT @joestringer ?

Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: Ah I think I misread the configurations you mentioned -- I've adjusted to only check native routing and tunneling for now

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml]: I agree with the sentiment that we may not need to run this test on every PR - most code changes would not impact this test. Currently we largely only have tests that run when PRs are posted and tests that are run upon merge. However if something fails on merge then the triage process is largely ad-hoc for correlating the failure back to the PR that triggers the failure.

The test run currently on this PR seems to take <20m, while for smoke tests we'd ideally be <10m (on this PR: 6-7m). So maybe it's not a bad idea to switch this over to being part of the `/test` jobs after all.

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml]: @marseel in tunneling mode, the local Pod's identity is put into packets based on the labels either provided from CNI ADD or maybe fetched from the Pod resource. The agent then uses this information to create the CEP object regardless. Separately that agent configures the datapath to include the source identity in tunnel headers, which is transmitted to a destination that may use it to make ingress policy decisions. It seems to me like this is effectively determining that for testing ingress network policy, CEP/CES may either be the basis for the policy decision, or not at all used. On the other hand, for egress policy, we cannot receive the destination identity in tunnel packets, so this information is always based on ipcache and hence CEP/CES. So regardless of the configuration, if egress policy was involved then the migration should have an effect. This seems like it suggests that the different settings are not going to test anything different with respect to CEP/CES handling? Or you have in mind a non-policy related use of Security Identity?

I see that this is using `.github/actions/conn-disrupt-test/action.yaml`, which at a glance seems like it doesn't involve policy at all as part of the disruption validation (https://github.com/cilium/cilium-cli/blob/81fe8664a5c63037157ed104bde55d903215c857/connectivity/check/deployment.go#L423), so I think the point is also moot about the identity being used for the forwarding?

To consider other cases, I think that the way we route traffic might be more relevant. IIRC that is tied into the IPAM settings. For instance I recall there is a mode where we rely on CEP/CES updates to understand which IPs reside on which nodes. In that scenario, depending on whether the CEP/CES information is propagated correctly and consistently, maybe the test could detect a drop in connectivity following a migration?

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [inline .github/workflows/tests-ces-migrate.yaml]: Is this still needed? The matrix doesn't specify it.

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml]: ^ I think the main downside is that if we need to further change this workflow, it's simply going to be easier if it's based on `pull_request`. I'd like to get feedback from someone in @cilium/ci-structure on this, maybe @viktor-kurchenko or @nbusseneau could chime in too if they have thoughts on this.

File: .github/workflows/tests-ces-migrate.yaml
Comment (marseel) [reply .github/workflows/tests-ces-migrate.yaml]: > CEP/CES may either be the basis for the policy decision, or not at all used

you are right, lets stick with native-routing to cover the ingress case too once we add the ingress policy.

> which at a glance seems like it doesn't involve policy at all as part of the disruption validation

true, I think we should add it, regardless of CES.

Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml]: > true, I think we should add it, regardless of CES.

What is "it"? I think the use of conn-disrupt makes sense and we can always extend it to introduce a policy scenario down the road, unrelated to this PR.

Comment (joestringer) [reply .github/workflows/tests-ces-migrate.yaml]: > > CEP/CES may either be the basis for the policy decision, or not at all used
> 
> you are right, lets stick with native-routing to cover the ingress case too once we add the ingress policy.

So do you mean we should just have the tunnel=disabled case and remove the matrix, or do you think there are settings we should include in the matrix?

Comment (marseel) [reply .github/workflows/tests-ces-migrate.yaml]: By "it" I meant including policies in conn-disrupt test.

I was refering to having test only for tunnel=disabled.

File: .github/workflows/tests-ces-migrate.yaml
Comment (joestringer) [inline .github/workflows/tests-ces-migrate.yaml]: Per the conclusions in [this thread](https://github.com/cilium/cilium/pull/32268#discussion_r1619554035) I think we can drop this matrix option.

```suggestion
```

File: .github/workflows/tests-ces-migrate.yaml
Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: Thanks Joe and Marcel, this was informative :) I've updated the code to only run the native routing case now.

File: .github/workflows/tests-ces-migrate.yaml
Comment (nbusseneau) [reply .github/workflows/tests-ces-migrate.yaml]: `pull_request` is definitely fine here, just like for our smoke tests this does not require access to any external resources. However if it does take a long time and also does not need to run on every PR, it's indeed a resource hog.

Personally I would say if we can be confident that this workflow only needs to run when changes are made to specific files / directories, then we can keep it `pull_request`-based and do the filtering in workflow, but if we cannot and considering we don't want `pull_request`-based workflows to take long, then we should move to Ariane even if annoying.

Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: Thanks all, this was all very helpful and interesting to learn! :) I'm wondering if one of you knows if it is possible to filter this workflow / the scenarios in which you would want to run this test? 

And if not, will it be possible for me to change to an Ariane /test workflow from this PR (from an external fork) or should it be approached differently? I may still be a bit confused about the cause of the problem with testing using that method from earlier.

Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: cc: @thorn3r @joestringer for your thoughts? :)

Comment (thorn3r) [reply .github/workflows/tests-ces-migrate.yaml]: Since this workflow is specific to the migration from CEP to CES, it seems to me like `pull_request` + filtering would be ideal. I think looking for changes in `pkg/k8s/apis/cilium.io` and `operator/pkg/ciliumendpointslice` would give us enough coverage for the migration.

Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: Got it, thank you!

**Comment by jshr-w** [comment]
Thanks everyone for all the help and feedback :) I think I'm ready for another review now (cc @thorn3r)

File: .github/workflows/tests-ces-migrate.yaml
Comment (thorn3r) [inline .github/workflows/tests-ces-migrate.yaml]: Since we added the paths to trigger this workflow, I think we can drop this job since it will always pass

**Review by thorn3r — APPROVED** [review]
LGTM! :ship:

File: .github/workflows/tests-ces-migrate.yaml
Comment (jshr-w) [reply .github/workflows/tests-ces-migrate.yaml]: ah i missed it :o

**Comment by jshr-w** [comment]
@viktor-kurchenko @nbusseneau Hi, I think this PR is blocked on codeowner review from ci-structure, would you be able to help with this please?

**Review by viktor-kurchenko — APPROVED** [review]
LGTM, thank you!

File: .github/workflows/tests-ces-migrate.yaml
Comment (julianwiedmann) [inline .github/workflows/tests-ces-migrate.yaml:110]: I believe the usage of `conn-disrupt-test` needs adjustment for the changes in https://github.com/cilium/cilium/pull/32930.

Comment (giorio94) [reply .github/workflows/tests-ces-migrate.yaml:110]: Yep, just mentioned the same: https://cilium.slack.com/archives/C069SNDMLA2/p1718286768251579

Comment (giorio94) [reply .github/workflows/tests-ces-migrate.yaml:110]: The path filter should also be modified so that these changed are detected immediately.

**Pull request merged**
